### PR TITLE
support experimental option and buildkit

### PR DESCRIPTION
**Pull request by int-tt**
added a setting to enable buildkit which is Docker 's new Image builder.
In order to use buildkit you need to enable dockerd's experimental.

The following sample confirms the operation.
```
---
resource_types:
- name: buildkit-on
  type: docker-image
  privileged: true
  source:
    repository: inttt/docker-image-resource
    tag: latest

resources:
- name: git-resource
  type: git
  source:
    uri: https://github.com/int-tt/docker-image-resource
    branch: feat-support-buildkit

- name: docker-resource-image
  type: buildkit-on
  source:
    repository: inttt/docker-image-resource
    experimental: true

jobs:
- name: build-rootfs
  plan:
  - get: git-resource
  - put: docker-resource-image
    params:
      build: git-resource
      buildkit: true
```

This gives you the benefit of faster build by buildkit :)
https://github.com/moby/buildkit

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@int-tt Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/docker-image-resource&pullRequestId=248)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=concourse/docker-image-resource&pullRequestId=248) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@int-tt Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/docker-image-resource&pullRequestId=248)!

**Comment by bkonkle** [comment]
Yes, please! 😀

File: README.md [hunk 1]
```diff
@@ -94,6 +94,8 @@ Note: docker registry must be [v2](https://docs.docker.com/registry/spec/api/).
  * `max_concurrent_uploads`: *Optional.* Maximum concurrent uploads.
 
    Limits the number of concurrent upload threads.
+  
+  * `experimental`: *Optional.* Use Dockerd experimental option
```
Comment (vito) [inline README.md]:
```suggestion
  * `experimental`: *Optional.* Use Dockerd experimental option.
```

File: README.md [hunk 2]
```diff
@@ -259,6 +261,8 @@ version is the image's digest.
 
 * `target_name`: *Optional.*  Specify the name of the target build stage. 
   Only supported for multi-stage Docker builds
+* `buildkit`: *Optional.*  Default`false`. If true use　Buildkit.
```
Comment (vito) [inline README.md]:
```suggestion
* `buildkit`: *Optional.* Default `false`. If `true`, use Buildkit. 
```

File: README.md (hunk 1 above)
Comment (vito) [inline README.md]: Could this instead be implied by `buildkit: true`? Does enabling this on its own make any difference?

File: README.md (hunk 2 above)
Comment (vito) [reply README.md]: Also this should maybe be called `use_buildkit: true`

File: tests/out_test.go [hunk 3]
```diff
@@ -303,6 +303,22 @@ var _ = Describe("Out", func() {
 		})
 	})
 
+	Context("when configured with a experimental", func() {
```
Comment (vito) [inline tests/out_test.go]: looks like there's no test for the buildkit param - might need to modify the `docker` stub to dump env so the tests can check against it

**Comment by vito** [comment]
Hey, sorry for the mixed signals but now that I think about it, it might be best to just hold off on this in favor of pushing folks towards [`registry-image` resource](https://github.com/concourse/registry-image-resource) and [`builder` task](https://github.com/concourse/builder) (which actually uses `buildkit` by way of `img` already).

There's more info on this in the v5.0 release notes: https://concourse-ci.org/download.html#v500-note-8

We're aiming to deprecate this resource type in favor of those two in the future, so adding experimental features may not be wise at this point - we should just be keeping it stable while people transition.

Does that sound OK? 🙏

**Comment by vito** [comment]
Closing per my previous comment.

For anyone looking to use `buildkit`, have a look at [`vito/oci-build-task`](https://github.com/vito/oci-build-task)!

**Pull request closed without merging**